examples/erc1155/lib.rs

```diff
@@ -0,0 +1,827 @@
+                        )
+                    }
+                    Err(e) => {
+                        match e {
+                            ink_env::Error::CodeNotFound
+                            | ink_env::Error::NotCallable => {
+                                // Our recipient wasn't a smart contract, so there's nothing more for
+                                // us to do
+                                ink_env::debug_println!("Recipient at {:?} from is not a smart contract ({:?})", from, e);
+                            }
+                            _ => {
+                                // We got some sort of error from the call to our recipient smart
+                                // contract, and as such we must revert this call
+                                let msg = ink_prelude::format!(
+                                    "Got error \"{:?}\" while trying to call {:?}",
+                                    e,
+                                    from
+                                );
+                                ink_env::debug_println!("{}", &msg);
+                                panic!("{}", &msg)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    impl super::Erc1155 for Contract {
+        #[ink(message)]
+        fn safe_transfer_from(
+            &mut self,
+            from: AccountId,
+            to: AccountId,
+            token_id: TokenId,
+            value: Balance,
+            data: Vec<u8>,
+        ) -> Result<()> {
+            let caller = self.env().caller();
+            if caller != from {
+                ensure!(self.is_approved_for_all(from, caller), Error::NotApproved);
+            }
+
+            ensure!(to != AccountId::default(), Error::ZeroAddressTransfer);
+
+            let balance = self.balance_of(from, token_id);
+            ensure!(balance >= value, Error::InsufficientBalance);
+
+            self.perform_transfer(from, to, token_id, value);
+            self.transfer_acceptance_check(caller, from, to, token_id, value, data);
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        fn safe_batch_transfer_from(
+            &mut self,
+            from: AccountId,
+            to: AccountId,
+            token_ids: Vec<TokenId>,
+            values: Vec<Balance>,
+            data: Vec<u8>,
+        ) -> Result<()> {
+            let caller = self.env().caller();
+            if caller != from {
+                ensure!(self.is_approved_for_all(from, caller), Error::NotApproved);
+            }
+
+            ensure!(to != AccountId::default(), Error::ZeroAddressTransfer);
+            ensure!(!token_ids.is_empty(), Error::BatchTransferMismatch);
+            ensure!(
+                token_ids.len() == values.len(),
+                Error::BatchTransferMismatch,
+            );
+
+            let transfers: Vec<(_, _)> = token_ids.iter().zip(values.iter()).collect();
```

Comment:
I would try not to do unnecessary heap memory allocations even though it cleans up the code to some extend. Try to use `clone` instead on the iterator maybe.

Reply:
Sorry, I don't quite understand your suggestion here. How would cloning the vectors be cheaper than allocating a new vector in this case?

Reply:
You'd clone an iterator over the original `token_ids`. Since we are only operating on shared references I guess `Clone` and maybe even `Copy` should exist for the iterator.

Reply:
This doesn't make sense to me 😅 
Can you push some code with what you mean?

Reply:
I'm not convinced this is any cheaper in terms of allocations, but I've gone with your suggestion